From 07a48bac15e23bbdf3d6d71036d15418459d124f Mon Sep 17 00:00:00 2001
From: Talat Batheesh <talatb@mellanox.com>
Date: Thu, 14 Dec 2017 11:07:49 +0200
Subject: [PATCH] net/mlx5: Fix fixpoint divide exception in
 mlx5e_am_stats_compare
Patch-mainline: Never, different upstream patch not applicable for SLE15
References: bsc#1071289

LTC BugProxy reported the following

[ 5515.229642] Kernel BUG at 000003ff81101652 [verbose debug info unavailable]
[ 5515.229643] Kernel BUG at 000003ff81101652 [verbose debug info unavailable]
[ 5515.229664] fixpoint divide exception: 0009 ilc:2 [#1] SMP
[ 5515.229693] CPU: 6 PID: 0 Comm: swapper/6 Not tainted 4.12.14-4.2-default #1
[ 5515.229694] Hardware name: IBM 3906 M03 703 (LPAR)
[ 5515.229695] task: 00000000f955a000 task.stack: 00000000f9560000
[ 5515.229696] Krnl PSW : 0704e00180000000 000003ff81101652 (mlx5e_am_stats_compare+0x82/0xe0 [mlx5_core])
[ 5515.229764]            R:0 T:1 IO:1 EX:1 Key:0 M:1 W:0 P:0 AS:3 CC:2 PM:0 RI:0 EA:3
[ 5515.229766] Krnl GPRS: 0000000000000128 000003ff00000000 00000000f8dd3d14 00000000e35a1434
[ 5515.229767]            00000000000073a0 000000000000000a 000000000000149f 0000000000000128
[ 5515.229768]            00000000e35a1430 0000000000000003 000000000003cb81 00000000000073a0
[ 5515.229769]            00000000e35a4608 000003ff81119e50 000003ff811017d6 00000000f8dd3c68
[ 5515.229776] Krnl Code: 000003ff81101646: b9140044            lgfr    %r4,%r4
                          000003ff8110164a: b90400b4            lgr     %r11,%r4
                         #000003ff8110164e: b91d00a1            dsgfr   %r10,%r1
                         >000003ff81101652: ecb2001e0a7e        cij     %r11,10,2,3ff8110168e
                          000003ff81101658: 58103008            l       %r1,8(%r3)
                          000003ff8110165c: 58402008            l       %r4,8(%r2)
                          000003ff81101660: b9f91024            srk     %r2,%r4,%r1
                          000003ff81101664: 1022                lpr     %r2,%r2
[ 5515.229790] Call Trace:
[ 5515.229823] ([<000003ff81101874>] mlx5e_rx_am+0x114/0x348 [mlx5_core])
[ 5515.229837]  [<000003ff81101daa>] mlx5e_napi_poll+0x302/0x310 [mlx5_core]
[ 5515.229840]  [<00000000005de064>] net_rx_action+0x134/0x3e8
[ 5515.229844]  [<0000000000713278>] __do_softirq+0x120/0x348
[ 5515.229847]  [<00000000001472c2>] irq_exit+0xba/0xd0
[ 5515.229850]  [<000000000010ba02>] do_IRQ+0x6a/0x88
[ 5515.229851]  [<00000000007126d4>] io_int_handler+0xfc/0x250
[ 5515.229853]  [<0000000000102ba6>] enabled_wait+0x4e/0xe0
[ 5515.229853] ([<0000000000000000>]           (null))
[ 5515.229855]  [<0000000000102ee2>] arch_cpu_idle+0x32/0x48
[ 5515.229857]  [<000000000018fca0>] do_idle+0xe8/0x1a0
[ 5515.229858]  [<000000000018ff2e>] cpu_startup_entry+0x3e/0x48
[ 5515.229861]  [<00000000001151ec>] smp_start_secondary+0xf4/0x108
[ 5515.229863]  [<0000000000712c6e>] restart_int_handler+0x62/0x78
[ 5515.229863]  [<0000000000000000>]           (null)
[ 5515.229864] Last Breaking-Event-Address:
[ 5515.229877]  [<000003ff81101614>] mlx5e_am_stats_compare+0x44/0xe0 [mlx5_core]
[ 5515.229878]
[ 5515.229879] fixpoint divide exception: 0009 ilc:2 [#2]
[ 5515.229880] Kernel panic - not syncing: Fatal exception in interrupt

This devision by ziro happen since the ppms become zero after cable plug/unplg.

This patch prevent the division by ziro by protect the ppms and epms.

Signed-off-by: Talat Batheesh <talatb@mellanox.com>
Acked-by: Talat Batheesh <tbogendoerfer@suse.de>
---
 drivers/net/ethernet/mellanox/mlx5/core/en_rx_am.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/net/ethernet/mellanox/mlx5/core/en_rx_am.c b/drivers/net/ethernet/mellanox/mlx5/core/en_rx_am.c
index acf32fe..c937f22 100644
--- a/drivers/net/ethernet/mellanox/mlx5/core/en_rx_am.c
+++ b/drivers/net/ethernet/mellanox/mlx5/core/en_rx_am.c
@@ -189,6 +189,7 @@ static void mlx5e_am_exit_parking(struct mlx5e_rx_am *am)
 static int mlx5e_am_stats_compare(struct mlx5e_rx_am_stats *curr,
 				  struct mlx5e_rx_am_stats *prev)
 {
+
 	if (!prev->bpms)
 		return curr->bpms ? MLX5E_AM_STATS_BETTER :
 				    MLX5E_AM_STATS_SAME;
@@ -197,9 +198,16 @@ static int mlx5e_am_stats_compare(struct mlx5e_rx_am_stats *curr,
 		return (curr->bpms > prev->bpms) ? MLX5E_AM_STATS_BETTER :
 						   MLX5E_AM_STATS_WORSE;
 
+	if (!prev->ppms)
+		return curr->ppms ? MLX5E_AM_STATS_BETTER :
+		MLX5E_AM_STATS_SAME;
+
 	if (IS_SIGNIFICANT_DIFF(curr->ppms, prev->ppms))
 		return (curr->ppms > prev->ppms) ? MLX5E_AM_STATS_BETTER :
 						   MLX5E_AM_STATS_WORSE;
+	if (!prev->epms)
+		return curr->epms ? MLX5E_AM_STATS_WORSE :
+				    MLX5E_AM_STATS_SAME;
 
 	if (IS_SIGNIFICANT_DIFF(curr->epms, prev->epms))
 		return (curr->epms < prev->epms) ? MLX5E_AM_STATS_BETTER :
-- 
2.5.0

